### ISSUE TITLE
More efficient JPG and PNG handling in PDF export

### ISSUE THREAD

**Issue by TS53** [typst/typst#5278]
### Description

JPG images can get included without reencoding the image when exporting to PDF.

We can see this with the following latex code and [this](https://polona.pl/item-view/dac0ab7a-43b6-48e5-8492-a38e0b744cde?page=0) file (filesize=8.4 mb):

```latex
\documentclass{article}

\usepackage{graphicx}

\begin{document}
\includegraphics[width=4cm]{map.jpg}
\end{document}
```

The resulting PDF compiles fast and has filesize 8.41 mb.

With typst, it seems the JPG is decoded and then reencoded, which is slower and makes the filesize bigger:

```typst
#image(width: 4cm, "map.jpg")
```

The resulting PDF has filesize 9.29 mb.

Note: for PNG also a [variant](https://tex.stackexchange.com/questions/39929/fast-png-embedding-using-pdflatex) exists that can get copied into the document without reencoding.

### Use Case

When exporting to PDF and having JPG images included:
- smaller binary size 
- faster compilation times
- probably better image quality (original is kept)

**Comment by TS53** [comment]
I think the `RasterImage` is already in a decoded format. If true, then this would mean that the image is at this point already stored into memory which will require some CPU time for decompression and is problematic for big images because it will at least temporarily load the full image into RAM like shown [here](https://github.com/typst/typst/discussions/5279).
Of course this cannot be prevented for all image formats.
From my understanding, JPG can just get copied into the final file without decoding the image and loading its contents into memory.

**Comment by laurmaedje** [comment]
Yes, JPEG could be handled more efficiently and even for PNG sometimes reuse is possible (though we'd more low-level access to the PNG). I had started to work on that a while ago, but didn't finish.

**Comment by ghost** [comment]
When compiled from main branch, it works now. The resulting PDF is only 2 kb bigger.
With PNG it is still slow, but I think this is expected because only few PNG files can be used directly.

Also, for JPG it seems the file is still opened and maybe decompressed which still slows down a bit.
A workaround is to include the JPG in a single PDF and then load this PDF with the single image instead of the image (`#image("pdfwithimg.pdf")`).
This will have the expected fast runtime.

**Comment by laurmaedje** [comment]
The JPG is decoded because an invalid JPG should be a compile error instead of just generating an invalid PDF.

**Comment by rugk** [comment]
Hm is this really fixed? I still see big sizes if I include an `image` in a `grid` with `auto` size? Or should I open a new issue about this?

**Comment by laurmaedje** [comment]
@rugk It's a little hard to know without further information. If you think there's still a bug, you can open a new issue with more details and we can investigate there.
